### PR TITLE
EOS:16437 - Add new python provisioning script '_s3_setup' with 'post_install' and 'cleanup'

### PR DESCRIPTION
**Pull request by amitwac2608**
We need to replace present 's3_setup', which is a shell based script, with a python based script.

As of now keeping the new python provisioning script name as '_s3_setup', but its name will
be replaces as soon as we EOL the shell script with same name.

Signed-off-by: Amit Kumar <amit.kumar@seagate.com>

### CODE REVIEW

**Review by pratyush-seagate — COMMENTED** [review]
Please look into the mini-provisioning failure.

**Comment by pratyush-seagate** [comment]
Since previous s3setup is being removed from rpm, the spec files needs to be updated accordingly.

**Comment by amitwac2608** [comment]
> Since previous s3setup is being removed from rpm, the spec files needs to be updated accordingly.

I am planning to remove the s3setup from rpm and repository, once all of "_s3_setup" is complete.

File: scripts/provisioning/_s3_setup [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+CONFSTORE_OPENLDAP_CONST_KEY = 'openldap'
+
+CONFSTORE_LDAPADMIN_USER_KEY = 'openldap>sgiam>user'
+CONFSTORE_LDAPADMIN_PASSWD_KEY = 'openldap>sgiam>secret'
+
+#CONFSTORE_ROOTDN_USER_KEY = 'openldap>root>user'
+#CONFSTORE_ROOTDN_PASSWD_KEY = 'openldap>root>secret'
+
+class Cmd:
```
Comment (bkirunge7) [inline scripts/provisioning/_s3_setup:51]: can we move these classes to different files, we can keep only orchestration part in this file

File: scripts/provisioning/_s3_setup [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+      self.delete_ldap_connection()
+
+      # remove configurations
+      if ret_val == 0:
+        sys.stdout.write(f'deleting configuration files\n')
+        ret_val = self.delete_config_files()
+
+      return ret_val
+
+  def create_ldap_connection(self):
```
Comment (bkirunge7) [inline scripts/provisioning/_s3_setup:244]: is this part of a class

File: scripts/provisioning/_s3_setup (hunk 1 above)
Comment (amitwac2608) [reply scripts/provisioning/_s3_setup]: done

File: scripts/provisioning/_s3_setup (hunk 2 above)
Comment (amitwac2608) [reply scripts/provisioning/_s3_setup]: yes.

File: scripts/provisioning/s3_setup_init.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+import sys
+
+from s3_setup_cmd import Cmd
+
+class InitCmd(Cmd):
```
Comment (bkirunge7) [inline scripts/provisioning/s3_setup_init.py:25]: lets match file name with class name initcmd.py,

File: scripts/provisioning/s3_setup_constants.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+"""Constants"""
+LDAP_USER = "cn={},dc=seagate,dc=com"
+LDAP_URL = "ldapi:///"
+
+CONFSTORE_OPENLDAP_CONST_KEY = 'openldap'
```
Comment (bkirunge7) [inline scripts/provisioning/s3_setup_constants.py:25]: can these come from config file

Comment (amitwac2608) [reply scripts/provisioning/s3_setup_constants.py:25]: this is constants file.. you mean should we make this file a config file ?

Or some other already existing file ?

File: scripts/provisioning/s3_setup_init.py (hunk 3 above)
Comment (amitwac2608) [reply scripts/provisioning/s3_setup_init.py]: done

File: scripts/provisioning/cleanupcmd.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    # 1. Check if the account exists or not
+    # 2. Delete the account if it exists, else do nothing
+    self.create_ldap_connection()
+
+    for acc in self.account_cleanup_dict.keys():
+      if(self.if_ldap_account_exists(acc)):
+        sys.stdout.write(f'deleting ldap account: {acc}\n')
+        try:
+          self.delete_ldap_account(acc)
+          self.delete_ldap_connection()
```
Comment (amitwac2608) [inline scripts/provisioning/cleanupcmd.py:69]: connection should be closed after looping over all accounts.

File: scripts/provisioning/cleanupcmd.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        except Exception as e:
+          raise e
+      else:
+        sys.stdout.write(f'INFO: ldap account: {acc} does not exist, nothing to do.\n')
+
+    # remove configurations
+    try:
+      sys.stdout.write(f'deleting configuration files\n')
+      self.delete_config_files()
```
Comment (amitwac2608) [inline scripts/provisioning/cleanupcmd.py:78]: comment out this call, as this seems to be part of upgrade, which is not clear yet..

**Comment by gowthamchinna** [comment]
test this please

**Comment by gowthamchinna** [comment]
run mini-provisioning

File: scripts/provisioning/s3_setup_constants.py (hunk 4 above)
Comment (amitwac2608) [reply scripts/provisioning/s3_setup_constants.py]: I have changed this constant file to a yaml config file.

File: scripts/provisioning/cleanupcmd.py (hunk 5 above)
Comment (amitwac2608) [reply scripts/provisioning/cleanupcmd.py]: done

File: scripts/provisioning/cleanupcmd.py (hunk 6 above)
Comment (amitwac2608) [reply scripts/provisioning/cleanupcmd.py]: done

**Review by bkirunge7 — APPROVED** [review]

**Comment by amitwac2608** [comment]
run mini-provisioning

**Comment by amitwac2608** [comment]
there are 7 codacy issues, all are same:

image


this issue: D203 is in conflict with D211, which says there should not be any blank line. I found this as existing issue :

https://github.com/PyCQA/pydocstyle/issues/141

We should also disable D203 for our project.

**Comment by AjinkyaDhumal** [comment]
@nileshgovande please review this once.

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1

File: scripts/provisioning/cleanupcmd.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  ldap_cn = None
+
+  def __init__(self, config: str):
+    """Constructor."""
+    try:
+      super(CleanupCmd, self).__init__(config)
+
+      self.read_ldap_credentials()
+
+      with open(self.s3_prov_config) as provconfig:
```
Comment (nileshgovande) [inline scripts/provisioning/cleanupcmd.py:50]: Use confstore

File: scripts/provisioning/cleanupcmd.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+  def create_ldap_connection(self):
+    """Open ldap connection."""
+    self.ldap_conn = ldap.initialize(self.ldap_url)
+
+    self.ldap_conn.protocol_version = ldap.VERSION3
+    self.ldap_conn.set_option(ldap.OPT_REFERRALS, 0)
+    self.ldap_conn.simple_bind_s(self.ldap_cn.format(self.ldap_user), self.ldap_passwd)
+
+  def delete_ldap_connection(self):
```
Comment (nileshgovande) [inline scripts/provisioning/cleanupcmd.py:95]: Make this PIC

File: scripts/provisioning/setupcmd.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+      s3cipher_obj = CortxS3Cipher(None, False, 0, prov_config_yaml['CONFSTORE_OPENLDAP_CONST_KEY'])
+      cipher_key = s3cipher_obj.generate_key()
+
+      encrypted_ldapadmin_pass = self.s3confstore.get_config(prov_config_yaml['CONFSTORE_LDAPADMIN_PASSWD_KEY'])
+      self.ldap_passwd = s3cipher_obj.decrypt(cipher_key, encrypted_ldapadmin_pass)
+
+      self.ldap_user = self.s3confstore.get_config(prov_config_yaml['CONFSTORE_LDAPADMIN_USER_KEY'])
+    except IOError as ioe:
+      sys.stderr.write(f'failed to open config file: {self.s3_prov_config}, err: {ioe}\n')
```
Comment (nileshgovande) [inline scripts/provisioning/setupcmd.py:68]: Use confstore

**Review by nileshgovande — APPROVED** [review]

**Pull request merged**